File: packages/cpp-sequential/src/SHTSimple.cpp
#include "../include/SHTSimple.h"
#include <cmath>
#include <iostream>

SHTResults SHTSimple(const std::vector<uint8_t> binaryImage,
                     const SHTOptions options) {
  std::vector<SHTResult> results;
  uint32_t width = options.width;
  uint32_t height = binaryImage.size() / width;

  // Defaults handled in the structure definitions and/or bindings
  float samplingRho = options.sampling.rho;
  float samplingTheta = options.sampling.theta;

  uint32_t hsWidth = std::trunc(360 * samplingTheta);
  uint32_t hsHeight =
      std::trunc(std::sqrt(width * width + height * height) * samplingRho);

  std::vector<uint32_t> houghSpace(hsWidth * hsHeight);

  uint32_t maxValue = 0;
  float samplingThetaRad = M_PI / 180. / samplingTheta;

  for (uint32_t y = 0; y < height; y++)
    for (uint32_t x = 0; x < width; x++)
      if (binaryImage[y * width + x] == 1)
        for (uint32_t hx = 0; hx < hsWidth; hx++) {
          float hTheta = hx * samplingThetaRad;
          float ySpace = x * std::cos(hTheta) + y * std::sin(hTheta);

          if (ySpace >= 0) {
            uint32_t offset =
                ((uint32_t)std::trunc(ySpace * samplingRho)) * hsWidth + hx;
            maxValue = std::max(maxValue, ++houghSpace[offset]);
          }
        }

  for (uint32_t hy = 0; hy < hsHeight; hy++)
    for (uint32_t hx = 0; hx < hsWidth; hx++) {
      uint32_t offset = hy * hsWidth + hx;
      if (houghSpace[offset] / (float)maxValue > options.votingThreshold) {
        results.push_back({
            hy / samplingRho,   // rho
            hx / samplingTheta, // theta
        });
      }
    }

  if (!options.returnHSpace)
    houghSpace.clear();

  HSpace hSpace{houghSpace, hsWidth};
  return {results, hSpace};
}

File: packages/js-sequential/src/SHTSimple.ts
import { SHT, SHTOptions, SHTResult, HTResults } from "meta";

const SHTSimple: SHT = function (
  binaryImage: Uint8Array,
  options: SHTOptions
): HTResults<SHTResult> {
  const results: SHTResult[] = [];
  const width = options.width;
  const height = binaryImage.length / width;

  // Defaults
  const sampling = Object.assign({ rho: 1, theta: 1 }, options.sampling);
  const votingThreshold = options.votingThreshold || 0.75;
  const samplingRho = sampling.rho;
  const samplingTheta = sampling.theta;
  const samplingThetaRad = Math.PI / 180 / samplingTheta;

  const hsWidth = Math.trunc(360 * samplingTheta);
  const hsHeight = Math.trunc(
    Math.sqrt(width ** 2 + height ** 2) * samplingRho
  );
  const houghSpace = new Uint32Array(hsWidth * hsHeight);

  let maxValue = 0;

  for (let y = 0; y < height; y++)
    for (let x = 0; x < width; x++)
      if (binaryImage[y * width + x] === 1)
        for (let hx = 0; hx < hsWidth; hx++) {
          const hTheta = hx * samplingThetaRad;
          const ySpace = x * Math.cos(hTheta) + y * Math.sin(hTheta);

          if (ySpace >= 0) {
            const offset = ((ySpace * samplingRho) << 0) * hsWidth + hx;
            maxValue =
              maxValue < ++houghSpace[offset] ? houghSpace[offset] : maxValue;
          }
        }

  for (let hy = 0; hy < hsHeight; hy++)
    for (let hx = 0; hx < hsWidth; hx++) {
      const offset = hy * hsWidth + hx;
      if (houghSpace[offset] / maxValue > votingThreshold) {
        results.push({
          rho: hy / samplingRho,
          theta: hx / samplingTheta,
        });
      }
    }

  return {
    results,
    hSpace: {
      data: houghSpace,
      width: hsWidth,
    },
  };
};

export { SHTSimple };
